[tune] Add test for multi-tenancy workaround and documentation to FAQ (#32560)
ray-project/ray: doc/source/tune/faq.rst

    The `TUNE_ORIG_WORKING_DIR` environment variable was the original workaround for
    accessing paths relative to the original working directory. This environment
    variable is deprecated, and the `chdir_to_trial_dir` flag described above should be
    used instead.

.. _tune-multi-tenancy:

> How can I run multiple Ray Tune jobs on the same cluster at the same time (multi tenancy)?

Reviewer: ```suggestion
How can I run multiple Ray Tune jobs on the same cluster at the same time (multi-tenancy)?
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Running multiple Ray Tune runs on the same cluster at the same
time is not officially supported. We do not test this workflow and we recommend
using a separate cluster for each tuning job.

The reasons for this are:

1. When multiple Ray Tune jobs run at the same time, they compete for resources.
   One job could run all its trials at the same time, while the other job waits
   for a long time until it gets resources to run the first trial.

> 2. If it is easy to start a new Ray cluster on your infrastructure, there is often

Reviewer: ```suggestion
2. If starting a new Ray cluster on your infrastructure is easy, there is often
```

Author: I think this changes the meaning - if starting many clusters is easy, that's what they should do!

Reviewer: Good catch. I do my best to understand the text and sometimes I mangle the meaning. Thanks for your patience!

>    no benefit from a cost perspective of running one big cluster versus multiple smaller

Reviewer: ```suggestion
   no cost benefit to running one big cluster.
```

>    clusters. For instance, running one cluster of 32 instances incurs virtually the same

Reviewer: ```suggestion
   cluster. For instance, running one cluster of 32 instances incurs almost the same
```

   cost as running 4 clusters with 8 instances each.
3. Concurrent jobs are harder to debug. If a trial of job A fills the disk,
   trials from job B on the same node are impacted. In practice, it's hard
   to reason about these conditions from the logs if something goes wrong.
4. Some internal implementations in Ray Tune assume that you only have one job
   running at a time. This can lead to conflicts.

The fourth reason is especially problematic when you run concurrent tuning jobs. For instance,
a symptom is when trials from job A use parameters specified in job B, leading to unexpected
results.

> If you run into this problem, please refer to

Reviewer 2: ```suggestion
Please refer to
```

    The `TUNE_ORIG_WORKING_DIR` environment variable was the original workaround for
    accessing paths relative to the original working directory. This environment
    variable is deprecated, and the `chdir_to_trial_dir` flag described above should be
    used instead.

.. _tune-multi-tenancy:

> How can I run multiple Ray Tune jobs on the same cluster at the same time (multi tenancy)?

Author: Happy to move it to the AIR FAQ once we have one. I've added a link from the AIR user guides overview page.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> First, please acknowledge that running multiple Ray Tune runs on the same cluster at the same

Reviewer: ```suggestion
Running multiple Ray Tune runs on the same cluster at the same
```

> time is not officially supported. This means we do not test this workflow and we recommend

Reviewer: ```suggestion
time is not officially supported. We do not test this workflow and we recommend
```

> using a separate cluster for every tuning job.

Reviewer: ```suggestion
using a separate cluster for each tuning job.
```

The reasons for this are:

1. When multiple Ray Tune jobs run at the same time, they compete for resources.
   One job could run all its trials at the same time, while the other job waits

>    for a long time until it gets resources to even run one trial.

Reviewer: ```suggestion
   for a long time until it gets resources to run the first trial.
```

2. Concurrent jobs are harder to debug. If a trial of job A fills the disk,

>    trials from job B on the same node will suffer from it. In practice it's hard

Reviewer: ```suggestion
   trials from job B on the same node are impacted. In practice it's hard
```

   to reason about these conditions from the logs if something goes wrong.

> 3. Some internal implementations in Ray Tune actually assume that you only have one job

Reviewer: ```suggestion
3. Some internal implementations in Ray Tune assume that you only have one job
```

>    running at the same time. This can lead to conflicts.

Reviewer: ```suggestion
   running at a time. This can lead to conflicts.
```

> Especially the third point is a common problem when running concurrent tuning jobs. Symptoms

Reviewer: ```suggestion
The third reason is especially problematic when you run concurrent tuning jobs. Symptoms
```

> are for instance that trials from job A use parameters specified in job B, leading to unexpected

Reviewer: ```suggestion
trials from job A use parameters specified in job B, lead to unexpected
```

results.

> The technical reason for this is that Ray Tune uses a cluster-global key-value store to store trainables.

Reviewer: ```suggestion
This occurs because Ray Tune uses a cluster-global key-value store to store trainables.
```

> If two tuning jobs use a trainable (or Ray AIR trainer) with the same name, the second job will overwrite

Reviewer: ```suggestion
If two tuning jobs use a trainable (or Ray AIR trainer) with the same name, the second job overwrites
```

> this trainable. Trials from the first job that are started after the second job started will then use

Reviewer: ```suggestion
the trainable. Trials from the first job that start after the second job starts will then use
```

> this overwritten trainable - leading to problems if captured variables are used (as is the case e.g. in

Reviewer: ```suggestion
the overwritten trainable, leading to problems if captured variables are used (as is the case in
```

most Ray AIR trainers).

Workaround
''''''''''

We are working on resolving the issue of conflicting trainables in concurrent tuning jobs. Until this
is resolved, the main workaround you can use is to give each trainable class a unique name.

For Ray AIR trainers this could look like this:

.. code-block:: python

    import uuid
    from ray.train.torch import TorchTrainer

>     TorchTrainer.__name__ = "TorchTrainer_" + uuid.uuid4().hex[:8]

Reviewer 3: Python question - at this point is `TorchTrainer` an isolated reference? I.e. will this name change leak if other places in the code import `TorchTrainer`?

Author: I believe it will update the name attribute globally.

    trainer = TorchTrainer(
        # ...

>     )

Reviewer 2: Can we instead just point to a github issue that describes this workaround? I don't want to commit this to the docs (it is too hacky).

Author: Ok, will update the PR

> Then the separate Ray Tune jobs will all have their own trainable which will not conflict with

Reviewer: ```suggestion
Ray Tune jobs have their own trainable and do not conflict with
```

> each other.

Reviewer: ```suggestion
others.
```

> The disadvantage of this approach is that resuming is a bit more complicated. If you want to resume

Reviewer: ```suggestion
This approach has the disadvantage of complicating the process of resuming. To resume
```

> a run, you will have to use the same unique trainable name again before calling ``Trainer.restore()``.

Reviewer: ```suggestion
a run, use the same unique trainable name again before calling ``Trainer.restore()``.
```

> This means you'll likely have to store the trainable name somewhere, as we're not updating this automatically

Reviewer: ```suggestion
You have to store the trainable name, because we're not updating it automatically
```
